mount_slash/mount_slash.8
.\" $Id$
.\" %GPL_START_LICENSE%
.\" ---------------------------------------------------------------------
.\" Copyright 2015-2016, Google, Inc.
.\" Copyright (c) 2008-2015, Pittsburgh Supercomputing Center (PSC).
.\" All rights reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or (at
.\" your option) any later version.
.\"
.\" This program is distributed WITHOUT ANY WARRANTY; without even the
.\" implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
.\" PURPOSE.  See the GNU General Public License contained in the file
.\" `COPYING-GPL' at the top of this distribution or at
.\" https://www.gnu.org/licenses/gpl-2.0.html for more details.
.\" ---------------------------------------------------------------------
.\" %END_LICENSE%
.\" %PFL_MODULES lnet pflenv %
.Dd February 29, 2016
.Dt MOUNT_SLASH 8
.ds volume PSC \- SLASH2 Administrator's Manual
.Os http://www.psc.edu/
.Sh NAME
.Nm mount_slash
.Nd mount a
.Tn SLASH2
file system
.Sh SYNOPSIS
.Nm mount_slash
.Op Fl dUV
.Op Fl o Ar mountopt
.Ar mountpoint
.Sh DESCRIPTION
The
.Nm
utility attachs a
.Tn SLASH2
file system under the specified
.Ar mountpoint .
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl d
Enable low-level file system debug messages.
.It Fl o Ar mountopt
Specify additional behavioral options.
The options are as follows:
.Bl -tag -width Ds
.It Ic acl
Enable
.Dq POSIX
access control list
.Pq ACL
support.
.It Ic ctlsock Ns = Ns Ar path
Specify an alternative path for the named socket through which
.Nm
operation can be controlled via
.Xr msctl 8 .
The following tokens are replaced in the file name specified:
.Pp
.Bl -tag -offset indent -width Ds -compact
.It Ic %h
the machine hostname
.It Ic %%
a literal
.Sq %
character
.El
.Pp
The default is
.Pa /var/run/mount_slash. Ns Ic %h Ns Pa .sock .
.It Ic datadir Ns = Ns Ar path
Specify an alternative directory where runtime configuration is
accessed.
Defaults to
.Pa /var/lib/slash .
.It Ic mapfile
Use the mapfile
.Pa /var/lib/slash/mapfile
.Pq or wherever the data directory is configured; see Ic datadir
for translating local user account IDs to global SLASH2 user IDs and
vice versa as necessary.
The concept of
.Dq global
simply means the user account IDs present on the MDS.
.Pp
The format of each line in this file is:
.Bd -literal -offset 4n
.Ic user local Ar uid Ic remote Ar uid
.Ed
.Pp
Group mappings may also specified:
.Bd -literal -offset 4n
.Ic group local-uids Ar uid ... Ic remote Ar gid
.Ed
.It Ic pagecache_maxsize Ns = Ns Ar size
Specify the maximum amount of memory to which the file data cache can
grow.
A good heuristic is 10% of available system memory but proper tuning
depends heavily on workload characteristics.
.Pp
Human-readable sizes are recognized
.Pq e.g. Li 4g for 4 GiB .
.Pp
Defaults to
.Li 512m .
.It Ic root_squash
Enable root squash.
Root squash mode disables traditional superuser access and instead
treats superuser like any other user.
Note that this is not a security feature per se but more of a guard
against accidental behavior.
.It Ic slcfg Ns = Ns Ar path
Specify an alternative configuration file.
Defaults to
.Pa /etc/slash/slcfg .
.It Ic statfs_pref_ios_only
When retrieving
.Xr statfs 2
information
.Pq e.g.\& via Xr df 1 ,
only include the configured preferred IOS instead of the aggregate sum
across all resources in the deployment.
Defaults to off.
.El
.It Fl U
Perform a
.Xr umount 8
at the location specified by
.Ar mountpoint
before attempting to bring up the file system.
.It Fl V
Print version and exit.
.El
.\" %PFL_INCLUDE $PFL_BASE/doc/env.mdoc {
.\"	subsys => {
.\"		bmap	=> "Block map structures",
.\"		fcmh	=> ".Tn FID\ncache members",
.\"		info	=> "General operational information",
.\"	},
.\"	lnets_note => <<'EOF',
.\"		.Pp
.\"		This is determined by default automatically from
.\"		.Xr slcfg 5
.\"		and the machine's networking configuration but can be
.\"		overridden if set specifically in
.\"		.Xr slcfg 5
.\"		or in the environment.
.\"		EOF
.\"	env => {
.\"		CTL_SOCK_FILE => "Path to control socket file\n.Pq see Ic ctlsock .",
.\"	},
.Sh ENVIRONMENT
.Bl -tag -width 3n
.It Ev CTL_SOCK_FILE
Path to control socket file
.Pq see Ic ctlsock .
.It Ev LNET_ACCEPT
Specify constraints on which connections are accepted.
May be
.Ic secure
for allowing connections only from
.Tn TCP
port 1023 and lower
.I all
to allow all, or
.Ic none
to disallow connections from user processes.
Defaults to
.Ic all .
.It Ev LNET_ACCEPT_BACKLOG
Specify the
.Xr listen 2
queue depth.
Defaults to 127.
.It Ev LNET_ACCEPT_PORT
Specify an alternate local
.Xr tcp 7
port for listening network sockets created by the Lustre networking
stack.
Defaults to 988.
.It Ev LNET_ACCEPT_TIMEOUT
Specify the number of seconds to wait for activity after a socket is
accepted.
Defaults to 5 seconds.
.It Ev LNET_IP2NETS
The list of
.Xr ip 4
networks and corresponding Lustre network names.
.It Ev LNET_NETWORKS
The list of networks and interfaces known to the Lustre networking
stack.
.Pp
This is determined by default automatically from
.Xr slcfg 5
and the machine's networking configuration but can be
overridden if set specifically in
.Xr slcfg 5
or in the environment.
.Pp
Examples:
.Bd -literal -offset 3n
tcp1(eth0),tcp1(eth1)
sdp4(ib0)
.Ed
.It Ev LNET_ROUTES
The list of Lustre routes for interconnecting networks.
.Pp
Examples:
.Bd -literal -offset 3n
tcp1 3@ptl1
tcp1 [2560,2563,2568,2571,2572,2575]@ptl0
.Ed
.It Ev PFL_SYSLOG_IDENT
Set to a custom value to pass as the
.Ar ident
parameter to
.Xr openlog 3
.Po used when
.Ev PFL_SYSLOG
is enabled
.Pc .
The following tokens are replaced in the value specified:
.Pp
.Bl -tag -offset 3n -width 6n -compact
.It Cm %n
the program (binary) name
.El
.It Ev PFL_SYSLOG_PIPE
Set to non-zero to launch a process that forwards all log messages
to the given shell pipeline, usually
.Dq logger .
.It Ev PSC_DUMPSTACK Pq debugging
When segmentation violations or fatal error conditions occur, try to
print a stack trace if this variable is defined.
.It Ev PSC_LOG_FILE
This path specifies the file name where log messages are written.
The following tokens are replaced in the file name specified:
.Pp
.Bl -tag -offset 3n -width 6n -compact
.It Cm %t
the number of seconds since the Epoch
.It Cm %%
a literal
.Sq %
character
.El
.It Ev PSC_LOG_FORMAT
This
.Xr printf 3 Ns -like
value is prepended to each log message reported.
The following specifiers are recognized:
.Pp
.Bl -tag -offset 3n -compact -width 6n
.It Ic B
source code file base name
.It Ic D Ns Aq Ar fmt
.Xr strftime 3 Ns -like
format of time
.It Ic F
source code function name
.It Ic f
source code file full path name
.It Ic H
full system host name
.It Ic h
short system host name
.It Ic I
.Xr pthread 3
.Tn ID
of thread which generated log message
.It Ic i
system
.Tn ID
of daemon thread which generated log message
.It Ic L
numeric level of log message importance
.It Ic l
source code file line number
.It Ic P
file system request's originating process'
.Tn PID
.It Ic N
daemon executable file base name
.It Ic n
symbolic thread name which generated log message
.It Ic r
.Tn MPI
rank of process
.It Ic s
seconds since Epoch
.Pq same as Li %D\*(Lt%s\*(Gt
.It Ic t
subsystem
.Tn ID
which generated log message
.It Ic T
subsystem name which generated log message
.It Ic U
file system request's originating process' user
.Tn ID
.It Ic u
microseconds into current second when generated
.It Ic X
file system request's originating process'
executable name
.El
.Pp
Examples:
.Bd -literal -offset 3n
[%D<%s>:%06u %n %F %l]
.Ed
.It Ev PSC_LOG_LEVEL
Each log message has an associated
.Dq level
numeric value which can be used to filter granularity of importance.
This environment variable specifies the minimum level of importance
necessary for messages to be reported.
.Pp
The following values are recognized:
.Pp
.Bl -tag -compact -offset 3n -width 13n
.It Cm 0 , none
No logging
.It Cm 1 , error
Recoverable failures
.It Cm 2 , warn
Something wrong which requires attention
.Pq default
.It Cm 3 , notice
Something unusual which recommends attention
.It Cm 4 , info
Informational messages
.It Cm 5 , diag
Diagnostics
.It Cm 6 , debug
Debugging messages
.It Cm 7 , vdebug
Verbose debugging messages
.It Cm 8 , trace , all
Full trace messages
.El
.It Ev PSC_LOG_LEVEL_ Ns Ar subsys
Like
.Ev PSC_LOG_LEVEL ,
but applies only to the given
.Ar subsys .
.Pp
.Ar subsys
may be one of the following:
.Pp
.Bl -tag -compact -offset 3n -width 13n
.It Cm bmap
Block map structures
.It Cm def
Default
.It Cm fcmh
.Tn FID
cache members
.It Cm info
General operational information
.It Cm mem
Memory allocations and releases
.El
.Pp
Examples:
.Bd -literal -offset 3n
PSC_LOG_LEVEL_mem=debug
.Ed
.It Ev PSC_MAX_NMEMMAPS
On Linux, specify the number of memory maps the process address space is
allowed to be divided into via
.Pa /proc/sys/vm/max_map_count .
Defaults to one million.
.It Ev PSC_SYSLOG
Set to non-zero to send all log messages to
.Xr syslog 3
.It Ev PSC_SYSLOG_ Ns Ar subsys
Set to non-zero to send log messages from the given
.Ar subsys
to
.Xr syslog 3
in addition to
.Em stderr .
.It Ev USOCK_CPORT
Specify an alternate remote/peer
.Xr tcp 7
port for connecting networking sockets.
Defaults to 988.
.It Ev USOCK_FAIR_LIMIT
Specify the number of packets that can be received or transmitted
without calling
.Xr poll 2 .
Defaults to one.
.It Ev USOCK_MIN_BULK
Specify the smallest bulk size permissible.
Defaults to 1024.
.It Ev USOCK_NPOLLTHREADS
Specify the number of threads to spawn to check and perform activity from
sockets.
Defaults to the minimum of the number of cores available on the system
or one quarter of the number of file descriptors available to the
process.
.It Ev USOCK_PEERTXCREDITS
Specify the number of concurrent sends to any single peer.
Defaults to 8.
.It Ev USOCK_POLL_TIMEOUT
Specify the number of seconds to wait for activity on a socket before
temporarily tending to other duties and trying again.
Defaults to one.
.It Ev USOCK_PORTPID
Specify whether to fabricate Lustre
.Tn PID
values from a connecting peer's originating
.Tn TCP
port instead of from the value specified by the peer during connect.
Defaults to off.
.It Ev USOCK_SOCKBUFSIZ
Specify an alternative value for the amount of data that can either be
queued awaiting tranmission or accumulate in a socket by the kernel
awaiting reception by a process.
Defaults to whatever value is initialized by the operating system.
.It Ev USOCK_SOCKNAGLE
Specify whether to enable the Nagle algorithm on sockets.
Defaults to disabling the Nagle algorithm.
.It Ev USOCK_TIMEOUT
Specify the number of seconds before no response on a socket triggers a
drop by the networking stack.
Defaults to 50 seconds.
.It Ev USOCK_TXCREDITS
Specify the total number of concurrent sends accumulated across all
peers.
Defaults to 256.
.El
.\" }%
.Sh FILES
.Bl -tag -width Pa -compact
.It Pa /etc/slash/slcfg
default
.Tn SLASH2
network configuration
.It Pa /var/lib/slash/authbuf.key
default inter-daemon communication key
.It Pa /var/lib/slash/mapfile
default UID mapping file
.It Xo
.Pa /var/run/mount_slash. Ns Ic %h Ns Pa .sock
.Xc
default
.Nm
control socket
.El
.Sh SEE ALSO
.Xr slcfg 5 ,
.Xr sladm 7 ,
.Xr msctl 8 ,
.Xr slashd 8 ,
.Xr sliod 8 ,
.Xr umount 8


mount_slash/mount_slash.sh
#!/usr/bin/env bash
# $Id$

prog=mount_wokfs
mod=slash2.so
ctl=msctl

PATH=$(dirname $0):$PATH
. pfl_daemon.sh

usage()
{
	echo "usage: $0 [-dgOTv] [-F filter] [-P deployment-profile] [instance]" >&2
	exit 1
}

bkav=("$@")
while getopts "dF:gOP:Tv" c; do
	case $c in
	d) nodaemonize=1	;;
	F) filter=$OPTARG	;;
	g) filter=mygdb		;;
	O) once=1		;;
	P) prof=$OPTARG		;;
	T) testmail=1		;;
	v) verbose=1		;;
	*) usage		;;
	esac
done
shift $(($OPTIND - 1))

xargs=()
apply_host_prefs "$@"

base=$dir/$prof.s2

preinit

: ${mp:=/$prof}
umount -l -f $mp 2>/dev/null
[ -d $mp ] || mkdir -p $mp

# Initialization/configuration
ulimit -n 100000
ulimit -c $((1024 * 1024 * 1024 * 50))
export LD_LIBRARY_PATH=/usr/local/lib
export PSC_SYSLOG_info=1
export PSC_LOG_LEVEL=${PSC_LOG_LEVEL:-notice}
export PSC_LOG_LEVEL_info=info
export PSC_LOG_FILE=${PSC_LOG_FILE:-$base/log/$host.$name/%t}
export PSC_LOG_FILE_LINK=$(dirname $PSC_LOG_FILE)/latest

type modprobe >/dev/null 2>&1 && modprobe fuse

xargs+=(datadir=$base/var)
xargs+=(slcfg=$base/slcfg)
opts=$(IFS=, ; echo "${xargs[*]}")

mod_dir=$(dirname $(which $prog))/../lib/wokfs

rundaemon $filter $prog -L "insert 0 $mod_dir/$mod $opts" -U $mp
